modules/gapi/src/compiler/passes/match_pattern_to_substitute.cpp

```diff
@@ -0,0 +1,106 @@
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "pattern_matching.hpp"
+
+#include <ade/util/zip_range.hpp>
+
+namespace cv { namespace gimpl {
+
+namespace {
+using Graph = GModel::Graph;
+using Metadata = typename Graph::CMetadataT;
+using VisitedMatchings = std::list<std::pair<ade::NodeHandle, ade::NodeHandle>>;
+
+// Returns true if two DATA nodes are semantically identical:
+//  - both nodes have the same GShape
+//  - both nodes have the same storage
+//
+// @param firstMeta - metadata of first
+// @param secondMeta - metadata of second
+bool compareDataNodes(const Metadata& firstMeta,
+                      const Metadata& secondMeta)
+{
+    GAPI_Assert(firstMeta.get<NodeType>().t == NodeType::DATA);
+    GAPI_Assert(firstMeta.get<NodeType>().t == secondMeta.get<NodeType>().t);
+
+    const auto& firstData = firstMeta.get<Data>();
+    const auto& secondData = secondMeta.get<Data>();
+    // compare shape
+    if (firstData.shape != secondData.shape) {
+        return false;
+    }
+    // compare storage
+    if (firstData.storage != secondData.storage) {
+        return false;
+    }
+
+    // FIXME: is it enough to only check shape && storage?
+    return true;
+}
+
+// Returns matched pairs of {pattern node, substitute node}
+SubgraphMatch::M matchDataNodes(const Graph& pattern,
+                                const Graph& substitute,
+                                const std::vector<ade::NodeHandle>& patternNodes,
+                                std::vector<ade::NodeHandle> substituteNodes)
+{
+    SubgraphMatch::M matched;
+    for (auto it : ade::util::zip(patternNodes, substituteNodes)) {
+        const auto& pNode = std::get<0>(it);
+        const auto& sNode = std::get<1>(it);
+        if (!compareDataNodes(pattern.metadata(pNode), substitute.metadata(sNode))) {
+            return {};
+        }
+        matched.insert({ pNode, sNode });
+    }
+    return matched;
+}
+}  // anonymous namespace
+
+SubgraphMatch matchPatternToSubstitute(const cv::gimpl::GModel::Graph& pattern,
+                                       const cv::gimpl::GModel::Graph& substitute,
+                                       const cv::gimpl::Protocol& patternP,
+                                       const cv::gimpl::Protocol& substituteP)
+{
+    //---------------------------------------------------------------
+    // Match data nodes which start and end our pattern and substitute
+    const auto& patternDataInputs = patternP.in_nhs;
+    const auto& patternDataOutputs = patternP.out_nhs;
+
+    const auto& substituteDataInputs = substituteP.in_nhs;
+    const auto& substituteDataOutputs = substituteP.out_nhs;
+
+    // if number of data nodes doesn't match, abort
+    if (patternDataInputs.size() != substituteDataInputs.size()
+        || patternDataOutputs.size() != substituteDataOutputs.size()) {
+        return {};
+    }
+
```

Comment:
And also if check on size is the check for validity of user-defined substitution graph for this pattern, so, I have no idea how can we reject this pair of graphs with existent logic:
![image](https://user-images.githubusercontent.com/15359579/63776141-70fbdc00-c8e9-11e9-9d37-c69b8577cc84.png)

I also have no idea how to check does user intent to substitute given pattern this way or make a mistake.
I mean I am not sure if there any sense to check user-defined substitution graph validity, may be we can do things simpler and always assume that substitution graph is valid. So check on sizes may become assert statement as extra check for sure.

Reply:
As a matter of fact, I use assert later on for such stuff. I see your point (I think) and I tend to agree, As to "does user intend to substitute given pattern this way or makes a mistake" -> I guess, the best we can do now is just imagine that the user is __always right__ (let the compilation stage e.g. `inferMeta` to verify that)